### PR TITLE
Feat/fri

### PR DESCRIPTION
**Pull request by ThomasPiellard**
A basic version of the [FRI](https://drops.dagstuhl.de/opus/volltexte/2018/9018/) protocol is now available, in ecc/curve/fr/fri.

The multiplicative version only has been implemented, it works on the prime-order subgroup of the corresponding curves.

The version is not tunable yet (only by hand), and should not be used unless careful choice of the parameters (e.g. rho, the number of rounds of verification) have been made.

It provides a working base for future work on FRI (possibly using ECFFT).

### API

example:

```
  // random polynomial
  p := randomPolynomial(uint64(size), s)
  
  // create an iop interface, corresponding here to the FRI multiplicative version, radix 2.
  // iop implements the Iopp interface.
  iop := RADIX_2_FRI.New(uint64(size), sha256.New())
  
  // The Iopp interface implements the BuildProofOfProximity method, to create
  // a proof of proximity.
  proof, err := iop.BuildProofOfProximity(p)
  if err != nil {
  t.Fatal(err)
  }
  
  // The Iopp interface implements the VerifyProofOfProximity method, verifying
  // a proof of proximity.
  err = iop.VerifyProofOfProximity(proof)
```

Internally, a proof of proximity is a sequence of interactions between a prover and a verifier (made non interactive using Fiat Shamir), where each interaction consists of a Merkle path proof verification, and a consistancy check between one round to  the next.

### CODE REVIEW

File: internal/generator/fri/template/fri.go.tmpl [hunk 1]
```diff
@@ -0,0 +1,700 @@
...
+	var res OpeningProof
+	res.merkleRoot, res.proofSet, res.index, res.numLeaves = tree.Prove()
+
+	// set the claimed value, which is the first entry of the Merkle proof
+	res.ClaimedValue.SetBytes(res.proofSet[0])
+
+	return res, nil
+}
+
+func checkRoots(a, b []byte) error {
```
Comment (gbotrel) [inline internal/generator/fri/template/fri.go.tmpl:271]: kill this, replace by bytes.Equal(..)

File: internal/generator/fri/template/fri.go.tmpl [hunk 2]
```diff
@@ -0,0 +1,700 @@
...
+
+	// we have the following system
+	// p₁(g²ⁱ)+gⁱp₂(g²ⁱ) = p(gⁱ)
+	// p₁(g²ⁱ)-gⁱp₂(g²ⁱ) = p(-gⁱ)
+	// we solve the system for p₁(g²ⁱ),p₂(g²ⁱ)
+	s := len(pSorted)
+	res := make([]fr.Element, s/2)
+
+	var p1, p2, twoInv, acc fr.Element
+	twoInv.SetUint64(2).Inverse(&twoInv)
```
Comment (gbotrel) [inline internal/generator/fri/template/fri.go.tmpl:342]: twoInv appears in multiple functions, maybe be worth it to initialize the constant as a global package-level variable ? (in a init() func)

File: internal/generator/fri/template/fri.go.tmpl [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+
+		// g <- g²
+		gInv.Square(&gInv)
+
+	}
+
+	// last round, provide the evaluation. The fully folded polynomial is of size rho. It should
+	// correspond to the evaluation of a polynomial of degree 1 on ρ points, so those points
+	// are supposed to be on a line.
+	res.evaluation = make([]fr.Element, rho)
```
Comment (gbotrel) [inline internal/generator/fri/template/fri.go.tmpl:437]: if `rho` is a constant, `evaluation` should be an array of size rho in the struct, not a slice

File: internal/generator/fri/template/fri.go.tmpl [hunk 4]
```diff
@@ -0,0 +1,700 @@
...
+		}
+		mr, proofSet, _, numLeaves := t.Prove()
+
+		// c denotes the entry that contains the full Merkle proof. The entry 1-c will
+		// only contain 2 elements, which are the neighbor point, and the hash of the
+		// first point. The remaining of the Merkle path is common to both the original
+		// point and its neighbor.
+		c := si[i] % 2
+		res.interactions[i][c] = partialMerkleProof{mr, proofSet, numLeaves}
+		res.interactions[i][1-c] = partialMerkleProof{
```
Comment (gbotrel) [inline internal/generator/fri/template/fri.go.tmpl:478]: `1-c` is never `< 0`?

File: internal/generator/fri/template/fri.test.go.tmpl [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+func TestFRI(t *testing.T) {
+
+	parameters := gopter.DefaultTestParameters()
+	parameters.MinSuccessfulTests = 10
+
+	properties := gopter.NewProperties(parameters)
+
+	size := 4096
```
Comment (gbotrel) [inline internal/generator/fri/template/fri.test.go.tmpl:54]: remark; shouldn't his vary during the tests? 

Similarly, `randomPolynomial` is not super random; (ie always the same form with coeff squared)

File: internal/generator/fri/template/fri.go.tmpl [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+	ErrProximityTestFolding = errors.New("one round of interaction failed")
+	ErrOddSize              = errors.New("the size should be even")
+	ErrMerkleRoot           = errors.New("merkle roots of the opening and the proof of proximity don't coincide")
+	ErrMerklePath           = errors.New("merkle path proof is wrong")
+	ErrRangePosition        = errors.New("the asked opening position is out of range")
+)
+
+const rho = 8
+
+var NbRounds = 1
```
Comment (gbotrel) [inline internal/generator/fri/template/fri.go.tmpl:26]: is it a constant? if it's not, should be made a parameter to the functions where it's needed.

File: internal/generator/fri/template/fri.go.tmpl [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+
+// var ErrorRate float32
```
Comment (gbotrel) [inline internal/generator/fri/template/fri.go.tmpl:28]: deadcode

File: internal/generator/fri/template/fri.go.tmpl (hunk 4 above)
Comment (ThomasPiellard) [reply internal/generator/fri/template/fri.go.tmpl:478]: no, here c is 0 or 1 here. The c-th entry contains the full Merkle path, the (1-c)-th the partial Merkle path.

**Review by gbotrel — APPROVED** [review]

**Comment by gbotrel** [comment]
@yelhousni can you review before merge?

File: internal/generator/fri/template/fri.test.go.tmpl (hunk 5 above)
Comment (gbotrel) [reply internal/generator/fri/template/fri.test.go.tmpl:54]: @ThomasPiellard

File: ecc/bn254/fr/mimc/mimc_test.go [hunk 8]
```diff
@@ -0,0 +1,60 @@
+package mimc
+
+// import (
+// 	"testing"
+// )
+
+// func TestMimc(t *testing.T) {
```
Comment (gbotrel) [inline ecc/bn254/fr/mimc/mimc_test.go:7]: @ThomasPiellard why is this commented / in the PR ?

**Pull request merged**